src/software/ai/navigator/obstacle/obstacle_factory_test.cpp

```diff
@@ -0,0 +1,124 @@
+#include "software/ai/navigator/obstacle/obstacle_factory.h"
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+
+#include "software/new_geom/circle.h"
+#include "software/new_geom/point.h"
+#include "software/new_geom/polygon.h"
+#include "software/new_geom/rectangle.h"
+#include "software/test_util/test_util.h"
+#include "software/world/robot.h"
+
+
+class ObstacleFactoryTest : public testing::Test
+{
+   public:
+    ObstacleFactoryTest()
+        : current_time(Timestamp::fromSeconds(123)),
+          obstacle_factory(std::make_shared<ObstacleFactoryConfig>())
+    {
+    }
+
+    ObstacleFactory obstacle_factory;
+    Timestamp current_time;
+};
+
+TEST_F(ObstacleFactoryTest, create_rectangle_obstacle)
+{
+    Rectangle rectangle(Point(1, 3), Point(5, 8));
+    Polygon expected(Rectangle(Point(.883, 2.883), Point(5.117, 8.117)));
+    Obstacle obstacle = obstacle_factory.createObstacleFromRectangle(rectangle);
+
+    auto polygon_obstacle = dynamic_cast<PolygonObstacle&>(*obstacle);
+    EXPECT_EQ(expected.getPoints(), polygon_obstacle.getPolygon().getPoints());
+}
+
+TEST_F(ObstacleFactoryTest, create_ball_obstacle)
+{
+    Point origin(2.5, 4);
+    Rectangle rectangle(Point(1, 3), Point(5, 8));
+    Circle expected(origin, 0.1985);
+    Obstacle obstacle = obstacle_factory.createBallObstacle(origin);
+
+    auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
+    EXPECT_EQ(expected, circle_obstacle.getCircle());
+}
+
+TEST_F(ObstacleFactoryTest, create_robot_obstacle)
+{
+    Point origin(2.5, 4);
+    Rectangle rectangle(Point(1, 3), Point(5, 8));
+    Circle expected(origin, 0.207);
+    Obstacle obstacle = obstacle_factory.createRobotObstacle(origin);
+
+    auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
+    EXPECT_EQ(expected, circle_obstacle.getCircle());
+}
+
+TEST_F(ObstacleFactoryTest, stationary_robot_obstacle)
+{
+    Point origin(2.3, 3);
+    Vector velocity(0.0, 0.0);
+    Angle orientation(Angle::fromRadians(2.2));
+    AngularVelocity angular_velocity(AngularVelocity::fromRadians(-0.6));
+    Circle expected(origin, 0.041569219381653054);
+    Robot robot = Robot(3, origin, velocity, orientation, angular_velocity, current_time);
+    Obstacle obstacle = obstacle_factory.createVelocityObstacleFromRobot(robot);
+
+    auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
```

Comment:
Since you're `dynamic_cast`-ing to a reference type, this will throw a `std::bad_cast` exception if it fails. In other areas where we do this we wrap the cast and assertions in a try-catch block that explicitly adds a failure if the cast fails. See the action tests for an example.

https://en.cppreference.com/w/cpp/language/dynamic_cast

Reply:
fixed it